src/main/scala/cromwell/engine/workflow/ValidateActor.scala

```diff
@@ -38,44 +49,88 @@ class ValidateActor(wdlSource: WdlSource, workflowInputs: Option[WdlJson], workf
     case ValidateWorkflow =>
       validateWorkflow(sender())
-      // NOTE: self shuts down when the parent PerRequest shuts down
-  }
-
-  private def validateInputs(namespaceWithWorkflow: NamespaceWithWorkflow, maybeWorkflowInputs: Option[WdlJson]): Future[Unit] = {
-    maybeWorkflowInputs match {
-      case Some(wi) =>
-        for {
-          inputs <- Future(wi.parseJson).map(_.asJsObject.fields)
-          coercedInputs <- Future.fromTry(namespaceWithWorkflow.coerceRawInputs(inputs))
-        } yield ()
-      case None => Future.successful(())
-    }
+    // NOTE: self shuts down when the parent PerRequest shuts down
   }
 
   private def validateWorkflow(sentBy: ActorRef): Unit = {
     logger.info(s"$tag for $sentBy")
-    val futureValidation: Future[Unit] = for {
+    val futureValidation: Future[ValidationResult] = (for {
       namespaceWithWorkflow <- Future(NamespaceWithWorkflow.load(wdlSource))
-      validatedInputs <- validateInputs(namespaceWithWorkflow, workflowInputs)
-      validatedRuntimeOptions <- Future.fromTry(TryUtil.sequence(namespaceWithWorkflow.workflow.calls map {
-        call => CromwellRuntimeAttributes.validateKeys(call.task.runtimeAttributes.attrs.keySet, CromwellBackend.backend.backendType)
-      }))
-    } yield ()
+      validatedInputs: Option[WorkflowCoercedInputs] <- Future(validateInputs(namespaceWithWorkflow))
+      validatedWorkflowOptions: Option[WorkflowOptions] <- Future(validateWorkflowOptions(CromwellBackend.backend()))
+      validatedRuntimeAttrs: Seq[Set[String]] <- Future(validateRuntimeAttributes(namespaceWithWorkflow))
+    } yield ValidationSuccess(namespaceWithWorkflow, validatedInputs, validatedWorkflowOptions, validatedRuntimeAttrs)) recover {
+      case reason: Throwable => ValidationFailure(reason)
+    }
+
+    //TODO: IMO, the responsibility for returning a status code should be a part of the client. This class should only return a validation result,
+    // and the callee should convert that to a status code or anything that it wants
 
     // Now validate that this Future completed successfully:
-    futureValidation onComplete {
-      case Success(_) =>
+    futureValidation map {
+      case _: ValidationSuccess =>
         logger.info(s"$tag success $sentBy")
         sentBy ! RequestComplete(
           StatusCodes.OK,
           APIResponse.success("Validation succeeded."))
 
-      case Failure(ex) =>
-        val messageOrBlank = Option(ex.getMessage).mkString
+      case ValidationFailure(reason) =>
+        val messageOrBlank = Option(reason.getMessage).mkString
         logger.info(s"$tag error $sentBy: $messageOrBlank")
         sentBy ! RequestComplete(
           StatusCodes.BadRequest,
-          APIResponse.fail(ex))
+          APIResponse.fail(reason))
+    }
+  }
+
+  //TODO: With PBE, this should be defined in the backend.
+  private def validateRuntimeAttributes(namespaceWithWorkflow: NamespaceWithWorkflow): Seq[Set[String]] = {
```

Comment:
At a high level I think these changes should be an evolutionary step in moving validation code from `WorkflowDescriptor` to here rather than adapting the existing code or writing new validations.  There would have to be some minor adjustments to account for the fact that the existing `WorkflowDescriptor` validations assume the workflow ID exists, but the code that's in `WorkflowDescriptor` I think better represents how the validations should work.